aiida_workgraph/orm/general_data.py

```diff
@@ -33,53 +37,53 @@ def __str__(self):
 
-        :return: a value
+        :return: The unpickled value.
         """
-        import cloudpickle as pickle
+        self._check_pickle_protocol()
+        return self._get_value_from_file()
 
-        def get_value_from_file(self):
-            filename = "value.pkl"
-            # Open a handle in binary read mode as the arrays are written as binary files as well
-            with self.base.repository.open(filename, mode="rb") as f:
-                return pickle.loads(f.read())  # pylint: disable=unexpected-keyword-arg
-
-        # Return with proper caching if the node is stored, otherwise always re-read from disk
-        return get_value_from_file(self)
+    def _get_value_from_file(self):
+        """Read the pickled value from file and return it."""
+        with self.base.repository.open(self.FILENAME, mode="rb") as f:
+            return cloudpickle.loads(f.read())  # Deserialize the value
 
     def set_value(self, value):
-        """Set the contents of this node.
+        """Set the contents of this node by pickling the provided value.
 
-        :param value: the value to set
+        :param value: The Python value to pickle and store.
         """
-        import cloudpickle as pickle
-        import sys
+        # Serialize the value and store it
+        serialized_value = cloudpickle.dumps(value)
+        self.base.repository.put_object_from_bytes(serialized_value, self.FILENAME)
 
-        self.base.repository.put_object_from_bytes(pickle.dumps(value), "value.pkl")
+        # Store relevant metadata
         python_version = f"{sys.version_info.major}.{sys.version_info.minor}"
         self.base.attributes.set("python_version", python_version)
+        self.base.attributes.set("serializer_module", cloudpickle.__name__)
+        self.base.attributes.set("serializer_version", cloudpickle.__version__)
+        self.base.attributes.set("pickle_protocol", cloudpickle.DEFAULT_PROTOCOL)
 
-    def _using_value_reference(self):
-        """This function tells the class if we are using a list reference.  This
-        means that calls to self.get_value return a reference rather than a copy
-        of the underlying list and therefore self.set_value need not be called.
-        This knwoledge is essential to make sure this class is performant.
-
-        Currently the implementation assumes that if the node needs to be
-        stored then it is using the attributes cache which is a reference.
+    def _check_pickle_protocol(self):
+        """Check if the stored pickle protocol matches the current environment.
 
-        :return: True if using self.get_value returns a reference to the
-            underlying sequence.  False otherwise.
-        :rtype: bool
+        Raises:
+            ImportError: If the pickle protocol does not match.
         """
-        return self.is_stored
+        stored_pickle_protocol = self.base.attributes.get("pickle_protocol")
+
+        if stored_pickle_protocol != cloudpickle.DEFAULT_PROTOCOL:
```

Comment:
Do you really need to catch this? I remember that python pickling automatically would catch this case, not sure about clouldpickle but I would assume yes

Reply:
You are right, the clouldpickle will handle this when load the data. I change it to catch the error when unpickling the data, see [2d49a9b](https://github.com/aiidateam/aiida-workgraph/pull/229/commits/2d49a9b82901f3d0c9674fabfd05534090bd1211)